lazer/publisher_sdk/proto/governance_instruction.proto

```diff
@@ -0,0 +1,343 @@
+    // governance instruction and will reject instructions with the same or smaller sequence no.
+    // Note that if instructions are received out of order, some of them may become permanently
+    // rejected (e.g. if instruction #3 has been successfully processed before instruction #2 was observed,
+    // #2 will always be rejected).
+    // Sequence numbers are assigned and tracked separately for each governance source.
+    optional uint32 governance_sequence_no = 4;
+}
+
+// Specifies which shards the governance instruction applies to.
+message ShardFilter {
+    // The instruction applies to the specified shards.
+    message ShardNames {
+        // Must not be empty.
+        repeated string shard_names = 1;
+    }
+    // The instruction applies to the specified shard groups.
+    message ShardGroups {
+        // Must not be empty.
+        repeated string shard_groups = 1;
+    }
+    // [required]
+    oneof filter {
+        // The instruction applies to all shards.
+        google.protobuf.Empty all_shards = 1;
+        ShardNames shard_names = 2;
+        ShardGroups shard_groups = 3;
+    }
+}
+
+// An item of a governance instruction.
+message GovernanceDirective {
+    // [required] Specifies which shards the governance instruction applies to.
+    // The instruction applies to each shard independently of other shards and may apply
+    // at a different time. The instruction may succeed on some shards and fail on other shards.
+    // Note that each shard has its own list of governance sources and permissions,
+    // and a `GovernanceInstruction` is issued by a single source, so multiple instructions
+    // from different sources may be needed to apply a change to multiple shards or shard groups.
+    optional ShardFilter shard_filter = 1;
+    // [required]
+    // Note: when adding a new variant here, update `Permissions` as well.
+    oneof action {
+        CreateShard create_shard = 101;
+        AddGovernanceSource add_governance_source = 102;
+        UpdateGovernanceSource update_governance_source = 103;
+        SetShardName set_shard_name = 104;
+        SetShardGroup set_shard_group = 105;
+        ResetLastSequenceNo reset_last_sequence_no = 106;
+        AddPublisher add_publisher = 107;
+        UpdatePublisher update_publisher = 108;
+        AddFeed add_feed = 109;
+        UpdateFeed update_feed = 110;
+    }
+}
+
+// Permissions granted to a governance source.
+// bool fields in this message are optional and default to false (no permission).
+message Permissions {
+    enum ShardAction {
+        // Required by protobuf. Instruction will be rejected if this value is encountered.
+        SHARD_ACTION_UNSPECIFIED = 0;
+        CREATE_SHARD = 101;
+        ADD_GOVERNANCE_SOURCE = 102;
+        UPDATE_GOVERNANCE_SOURCE = 103;
+        SET_SHARD_NAME = 104;
+        SET_SHARD_GROUP = 105;
+        RESET_LAST_SEQUENCE_NO = 106;
+        ADD_PUBLISHER = 107;
+        ADD_FEED = 109;
```

Comment:
108?

Reply:
These are just the same numbers as in `GovernanceDirective`. There is gap because there is no `UpdatePublisher = 108` here. I prefer to treat these values as opaque anyway.